sencha-workspace/SlateCSV/app/view/ImporterController.js

```diff
@@ -257,5 +264,69 @@ Ext.define('SlateCSV.view.ImporterController', {
         } else {
             previewGrid.setHidden(true);
         }
+    },
+
+    onImportButtonClick: function() {
+        var me = this,
+            view = me.getView(),
+            useFirstRowForColumnNames = view.getUseFirstRowForColumnNames(),
+            comboBoxes = me.getValidComboBoxes(),
+            csvData = view.getCsvData(),
+            cols = Ext.Object.getKeys(csvData[0]),
+            vtypes = Ext.form.VTypes,
+            validations = [],
+            validationWindow = view.getValidationWindow(),
+            failures = 0;
+
+        Ext.Array.each(comboBoxes, function(combo) {
+            field = combo.findRecordByValue(combo.getValue());
+
+            if (field.get('vtype') && vtypes[field.get('vtype')] ) {
+                validations.push({
+                    fieldName: field.get('fieldName'),
+                    fieldLabel: field.get('label'),
+                    dataIndex: combo.dataIndex,
+                    label: combo.getFieldLabel(),
+                    vtype: field.get('vtype'),
+                    vtext: (field.get('vtypeText') || vtypes[field.get('vtype') +'Text']),
+                    invalidRows: []
+                });
+            }
+        });
+
+        Ext.Array.each(csvData, function(row,index) {
+            Ext.Array.each(validations, function(v) {
```

Comment:
@BillClinton these need to be `for` loops. `Ext.Array.each` is massively slower than a for loop due to the overhead of declaring a function and shifting variable scope constantly. Using this type of loop is OK when having the closure around each iteration provides a functional benefit (i.e. you want to declare some variables and have their values stick for handlers created within. It's also OK when it's a short loop that won't be run frequently but you should generally err towards using `for` loops instead.

This spot is a good example of somewhere it's an absolute no-go to use functional iterators though. Not only is the csvData loop liable to have a lot of iterations, but there is _another_ loop embedded inside it, so each time it processes a row it will be declaring a new function and then shifting scope for each validator